### PR TITLE
Add mizar-pod-controller and Grpc client

### PR DESCRIPTION
**Pull request by Hong-Chang**
mizar-pod-controller is one of the mizar controllers that listen to cluster events and send corresponding info to mizar backend.

**What type of PR is this?**
/kind feature

### CODE REVIEW

File: cmd/kube-controller-manager/app/mizarcontrollers.go
Comment (h-w-chen) [inline cmd/kube-controller-manager/app/mizarcontrollers.go:27]: is there a reason we choose 4 here?

File: pkg/controller/mizarcontrollers/mizar-pod-controller.go
Comment (h-w-chen) [inline pkg/controller/mizarcontrollers/mizar-pod-controller.go:182]: Isuggest to add logic of processing potential errors

File: pkg/controller/mizarcontrollers/util.go
Comment (h-w-chen) [inline pkg/controller/mizarcontrollers/util.go:16]: this small file seems to able to incorporate in other files, as it only has very limited stuff.

**Review by h-w-chen — COMMENTED** [review]
Thanks for the effort of adding mizar pod controller skeletal code; good first move. 
code generally looks fine; left a few feedback. Will take a look again after they are addressed.

**Comment by chenqianfzh** [comment]
I have an overall suggestion. Can we change the name of controller to MizarPodController, instead of just pod controller?

Though there is no other controller watching pod now in Arktos, but it is likely there may be another watching-pod controller in the future, since pod is such a fundamental resource.

File: cmd/kube-controller-manager/app/mizarcontrollers.go
Comment (chenqianfzh) [inline cmd/kube-controller-manager/app/mizarcontrollers.go:35]: Is this controller just about mizar pods? If Mizar pods have their own label, I suggest we use a filtered informer here. That should benefit the performance, especially if the percentage of mizar pods is not high.

File: pkg/controller/mizarcontrollers/mizar-pod-controller.go
Comment (chenqianfzh) [inline pkg/controller/mizarcontrollers/mizar-pod-controller.go:184]: I think it is ideal to use "switch-case" here.

Besides, the handling of "delete" and non-recognized event-type are missing.

File: pkg/controller/mizarcontrollers/mizar-pod-controller.go
Comment (chenqianfzh) [inline pkg/controller/mizarcontrollers/mizar-pod-controller.go:39]: If you decide to follow my suggestion to use the term MizarPodController consistently, we need only one const here.

File: pkg/controller/mizarcontrollers/mizar-pod-controller.go
Comment (chenqianfzh) [reply pkg/controller/mizarcontrollers/mizar-pod-controller.go:182]: +1

Errors other than NotFound are not handled.

File: cmd/kube-controller-manager/app/controllermanager.go
Comment (Hong-Chang) [reply cmd/kube-controller-manager/app/controllermanager.go:425]: I'm implementing a mechanism that watching configMap, and start controller after configMap shows it's ready for starting controller.

Comment (Hong-Chang) [reply cmd/kube-controller-manager/app/controllermanager.go:425]: I think it's better to hold the review until I make new iteration. I'll close the pr for now.

File: pkg/controller/mizarcontrollers/mizar-pod-controller.go
Comment (Hong-Chang) [reply pkg/controller/mizarcontrollers/mizar-pod-controller.go:34]: Updated.

File: pkg/controller/mizarcontrollers/mizar-pod-controller.go
Comment (Hong-Chang) [reply pkg/controller/mizarcontrollers/mizar-pod-controller.go:59]: Updated.

File: pkg/controller/mizarcontrollers/mizar-pod-controller.go
Comment (Hong-Chang) [reply pkg/controller/mizarcontrollers/mizar-pod-controller.go:61]: I see other controllers are using their own eventBroadcaster. Some module has multiple controllers but they still don't share.

File: pkg/controller/mizarcontrollers/mizar-pod-controller.go
Comment (Hong-Chang) [reply pkg/controller/mizarcontrollers/mizar-pod-controller.go:65]: Changed name from oc to c.

File: pkg/controller/mizarcontrollers/mizar-pod-controller.go
Comment (Hong-Chang) [reply pkg/controller/mizarcontrollers/mizar-pod-controller.go:94]: Inside WaitForCacheSync there is enough log for troubleshooting.

File: cmd/kube-controller-manager/app/mizarcontrollers.go
Comment (Hong-Chang) [reply cmd/kube-controller-manager/app/mizarcontrollers.go:27]: From other controllers, I see if pod/event happen frequently, the number is higher. For rarely happened controller, the number is 1. Considering pod is core object and its quantity could be huge, I think the number 4 is proper.

I made an update to don't put it in const. Because other mizar controllers are not necessary use 4.

File: pkg/controller/mizarcontrollers/mizar-pod-controller.go
Comment (Hong-Chang) [reply pkg/controller/mizarcontrollers/mizar-pod-controller.go:182]: Nice catch. Actually IsNotFound is not special here. Original I reserved it for deletion scenario but actually it doesn't apply. I changed code to return error when it happens. After error returns, it will be logged and retried.

File: pkg/controller/mizarcontrollers/util.go
Comment (Hong-Chang) [reply pkg/controller/mizarcontrollers/util.go:16]: There are some planned func coming. For example, I need a function to transfer from pod object to pod grpc contract object.

File: pkg/controller/mizarcontrollers/mizar-pod-controller.go
Comment (Hong-Chang) [reply pkg/controller/mizarcontrollers/mizar-pod-controller.go:184]: Updated to switch.
"delete" event planed to be handled in future release.

File: cmd/kube-controller-manager/app/mizarcontrollers.go
Comment (Hong-Chang) [reply cmd/kube-controller-manager/app/mizarcontrollers.go:35]: The pods are all scope. The name "mizar pod" just indicates this controller serves for mizar purpose.

**Comment by Hong-Chang** [comment]
> I have an overall suggestion. Can we fix the name of controller to MizarPodController? I see the term of "mizar-pod-controller" and "podController" are used interchangably.
> 
> Though there is no other controller watching pod now in Arktos, but it is likely there may be another watching-pod controller in the future, since pod is such a fundamental resource.
> 
> I suggest to make changes like renaming the function name from controllers.NewPodController to controllers.NewMizarPodController, and other things like that.

Good point. I did some renaming. Now Mizar and Pod are always together.

File: pkg/controller/mizarcontrollers/mizar-pod-controller.go
Comment (Hong-Chang) [reply pkg/controller/mizarcontrollers/mizar-pod-controller.go:39]: Updated.

File: pkg/controller/mizar/mizar-pod-controller.go
Comment (clu2xlu) [inline pkg/controller/mizar/mizar-pod-controller.go:51]: This part of the code I was not able to run on my end.  System complains unrecognized type  KeyWithEventType

File: pkg/controller/mizar/mizar-pod-controller.go
Comment (clu2xlu) [inline pkg/controller/mizar/mizar-pod-controller.go:80]: Could you change the naming here? It's kinda confusing

File: pkg/controller/mizar/util.go
Comment (clu2xlu) [inline pkg/controller/mizar/util.go:26]: I will move this part to my Service controller

File: cmd/kube-controller-manager/app/mizarcontrollers.go
Comment (chenqianfzh) [inline cmd/kube-controller-manager/app/mizarcontrollers.go:52]: let's not hard the number of worker as  4 here. Give it a variable, we may need to tune this in the future.

**Review by h-w-chen — COMMENTED** [review]
all of my comments had been addressed; the skeletal code of mizar-pod controller looks fine to me.

Can this PR refrain from doing more than skeletal controller? The grpc related code is better in its separate PR, and team members who are familiar with grpc can help to provide meaningful reviews.

[Hong] I agree with you. Our pr should be small and precise. We should split our task to small pieces and independent. 
In this case, it's not ideal because in the final stage of the release I jumped in the tasks and make them done. These tasks should have been done one by one but now I have to work on them all at once. The later task depend on prev task. And since I do them together (in the weekend of labor day), I have to send only one pr. I guess we don't have time to split them to 3 pr and review one by one.
I will make sure every part will be reviewed by you or Chen Qian.

File: pkg/controller/mizar/grpc-adaptor.go
Comment (chenqianfzh) [inline pkg/controller/mizar/grpc-adaptor.go:30]: Guess we can have simpler code by abstrction using func as a param. For example, for those Service related functions:

func GrpcServiceOperation(grpcHost string, service *v1.Service, operation func(context.Context, *BuiltinsServiceMessage) (*ReturnCode, erroRr) ) *ReturnCode {
	client, ctx, conn, 
        cancel, err := getGrpcClient(grpcHost)
	if err != nil {
		return getReturnCodeFromError(&err)
	}
	defer conn.Close()
	defer cancel()
	returnCode, err := operation(ctx, ConvertToServiceContract(service))
	if err != nil {
		return getReturnCodeFromError(&err)
	}
	return returnCode
}

func GrpcCreateService(grpcHost string, service *v1.Service) *ReturnCode {
       return GrpcServiceOperation(grpcHost, service, client.CreateService)
}

...

Comment (chenqianfzh) [reply pkg/controller/mizar/grpc-adaptor.go:30]: Actually, more abstraction can be done over the different types of resources. Yet it involves more type conversion and check, I had a brief look and it seems to me that the signature of those conversion funcs need to be updated to handler the possible conversion errors.

So I just stop here, you can play with it more to have even simpler code.

File: pkg/controller/mizar/grpc-adaptor.go
Comment (chenqianfzh) [inline pkg/controller/mizar/grpc-adaptor.go:264]: Why insecure?

File: pkg/controller/mizar/grpc-adaptor.go
Comment (chenqianfzh) [inline pkg/controller/mizar/grpc-adaptor.go:257]: I think HTTP Internal Error is more suitable in this case, which seems the usual behavior in K8s.

Comment (chenqianfzh) [reply pkg/controller/mizar/grpc-adaptor.go:257]: One thing more:  getReturnCodeFromError and getGrpcClient are very likely to be used a lot by other team members. Consider change them to public functions

File: pkg/controller/mizar/grpc-adaptor.go
Comment (chenqianfzh) [inline pkg/controller/mizar/grpc-adaptor.go:18]: fmt "fmt" -> "fmt"

File: pkg/controller/mizar/grpc-adaptor.go
Comment (chenqianfzh) [inline pkg/controller/mizar/grpc-adaptor.go:23]: v1 "k8s.io/api/core/v1" ->  "k8s.io/api/core/v1"

File: pkg/controller/mizar/mizar-pod-controller.go
Comment (chenqianfzh) [reply pkg/controller/mizar/mizar-pod-controller.go:80]: +1, a common func in this case is sync(). Does it fit?

File: pkg/controller/mizar/mizar-pod-controller.go
Comment (chenqianfzh) [inline pkg/controller/mizar/mizar-pod-controller.go:124]: Nit: make a new type, an enum of strings, for EventType,

File: pkg/controller/mizar/mizar-pod-controller.go
Comment (chenqianfzh) [inline pkg/controller/mizar/mizar-pod-controller.go:205]: Not sure whether it is OK to log the error as "temporary". When we say "temporary", it is usually an error that will go after retrial.

File: pkg/controller/mizar/mizar-starter-controller.go
Comment (chenqianfzh) [inline pkg/controller/mizar/mizar-starter-controller.go:160]: Don't check the key directly. Instead check the value of tenant/ns/name. The key format may be changed in the future.

File: cmd/kube-controller-manager/app/mizarcontrollers.go
Comment (Hong-Chang) [reply cmd/kube-controller-manager/app/mizarcontrollers.go:40]: It won't. Here in line 39 I have to declare controllerContext as interface{} instead of type ControllerContext. It's because the context will be pass into start controller then pass back when starting other controllers. Inside start controller I cannot import ControllerContext's package to avoid circle import.

File: pkg/controller/mizarcontrollers/mizar-pod-controller.go
Comment (Hong-Chang) [reply pkg/controller/mizarcontrollers/mizar-pod-controller.go]: Now delete event is being handled.

File: cmd/kube-controller-manager/app/mizarcontrollers.go
Comment (Hong-Chang) [reply cmd/kube-controller-manager/app/mizarcontrollers.go:52]: Updated. Added const.

File: pkg/controller/mizar/grpc-adaptor.go
Comment (Hong-Chang) [reply pkg/controller/mizar/grpc-adaptor.go:18]: Updated.

**Review by h-w-chen — APPROVED** [review]
approved based on arktos skeletal controller code.

**Review by Sindica — CHANGES_REQUESTED** [review]
I am not seeing any unit test or integration test code.

File: pkg/controller/mizar/mizar-starter-controller.go
Comment (chenqianfzh) [inline pkg/controller/mizar/mizar-starter-controller.go:160]: minor: "default" -> metav1.NamespaceDefault

**Review by chenqianfzh — APPROVED** [review]
LGTM.

Hong, can you update the thread that we synced offline? so we know they are addressed.

**Review by Hong-Chang — COMMENTED** [review]
> I am not seeing any unit test or integration test code.

I plan to write test code after this merged. Our release date is 9/15, and I originally didn't join the release. Later when I see there is risk on schedule so I joined the release to work on these code, and these code is like structural code that is dependent by other team members. So I need to code to be merged first to unblock other people. Then I'll write test code for it.

**Comment by Sindica** [comment]
> I am not seeing any unit test or integration test code.

Author got offline approved from Dr. Xiong.

**Comment by Hong-Chang** [comment]
> LGTM.
> 
> Hong, can you update the thread that we synced offline? so we know they are addressed.

Yes, we majorly discussed the refactoring of grpc-adaptor. Because currently it's not decided yet how other controller will use the grpc-adaptor, I will hold the refactoring until the code is accepted by other members. Qian agrees.

**Review by Sindica — APPROVED** [review]

**Pull request merged**
